py/bitbox02/communication/bitbox_api_protocol.py

```diff
@@ -180,19 +180,25 @@ def __init__(
         send_msg = noise.write_message()
         assert noise.handshake_finished
         pairing_code = base64.b32encode(noise.get_handshake_hash()).decode("ascii")
-        show_pairing_callback(
-            "{} {}\n{} {}".format(
-                pairing_code[:5], pairing_code[5:10], pairing_code[10:15], pairing_code[15:20]
-            )
-        )
         response = self._query(send_msg)
 
         # Can be set to False if the remote static pubkey was previously confirmed.
         pairing_verification_required_by_host = True
 
         pairing_verification_required_by_device = response == b"\x01"
         if pairing_verification_required_by_host or pairing_verification_required_by_device:
-            pairing_response = self._query(OP_I_CAN_HAS_PAIRIN_VERIFICASHUN)
+            cid = self._transport.generate_cid()
+            self._transport.write(OP_I_CAN_HAS_PAIRIN_VERIFICASHUN, HWW_CMD, cid)
+            client_response_success = show_pairing_callback(
+                "{} {}\n{} {}".format(
+                    pairing_code[:5], pairing_code[5:10], pairing_code[10:15], pairing_code[15:20]
+                )
+            )
+            if not client_response_success:
+                self.close()
+                raise Exception("pairing rejected by the user on client")
```

Comment:
you should raise a `PairingFailedException` or something like that, as that is an exception that is supposed to be handled by the client, e.g. by showing a (translated) error message. The other exceptions are kind of internal errors which cannot be handled specially.

Reply:
I would have guessed that if the client is capable of determining if the user rejected the pairing or not, there is no need for a specific exception.

Reply:
If I'd name the exceptions specifically, I would rather handle the user rejecting the pairing on the device properly, since that it is no error internal to the library, but the user executing an action. That would not be in scope of this commit though.

Reply:
What's the plan then, to silently ignore all exceptions raised here? If you plan on showing the error message in the exception, then treating this specially makes sense, as it is something we expect the user to see, versus the others which should not happen if everything is bug-free.

I guess for now it's good enough to just show the exception message (English) but the "proper" way is to have a translation entry.

Reply:
Handling a failed pairing is client side only, the device is not affected. Assume a MITM, you don't need to engage with it anymore when you reject a pairing. The app just needs to warn the user of what happened, or try again in case it was a transient error (which I hope will not happen again after we fix the one outstanding known bug related to this).